README.md
#codesort
A Python source-code organization tool.

The goal of this project is to create a tool that will take a python source code file (.py) and organize it to some standard by moving around class and function defintions, while still maintaining original functionality.

###The organization will be as follows:

>1. source file encoding (the "# -*- coding: utf-8 -*-" at the start of the file)
>2. module docstring
>3. \__future__ import statements
>4. import statements
>5. module constants
>6. classes
>7. module functions
>8. main() function
>9. toplevel module code
>10. if \__name__ == '\__main__' block


###Classes are to be organized in the following way:

>0. class docstring
>1. \__init__
>2. \__new__
>3. \__del__
>4. \__str__
>5. \__repr__
>6. \__cmp__
>7. \__hash__
>8. \__nonzero__
>9. \__unicode__
>10. \__getattr__
>11. \__setattr__
>12. \__delattr__
>13. \__getattribute__
>14. \__get__
>15. \__set__
>16. \__delete__
>17. \<private methods>
>18. \<public methods>


Private and private methods are organized alphabetically.

Module functions are organized alphabetically, except for the main() function, which is always at the end of the module functions


CHANGES.txt
v0.1.0, 2014-07-03 10:21 -- Initial release.


tests/test_codesort.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 18 16:26:08 2014

@name:          test_codesort.py
@vers:          0.1
@author:        dthor
@created:       Wed Jun 18 16:26:08 2014
@modified:      Wed Jun 18 16:26:08 2014
@descr:         Unit Testing for codesort.codesort module
"""

from __future__ import print_function, division
import unittest
import os.path
import codesort.codesort as codesort


class CodeSort(unittest.TestCase):
    """ Unit Testing """
    def setUp(self):
        self.path = r"X:\WinPython27\projects\codesort\tests\test_data"
        self.filenames = (("unsorted_1.py", "sorted_1.py"),
                          )
        self.known_values = []
        for test_file, ref_file in self.filenames:
            self.known_values.append((os.path.join(self.path, test_file),
                                      os.path.join(self.path, ref_file)))

    def test_known_values(self):
        """
        KVT for CodeSort: verify that the sorted test file matches
        with the reference file.
        """

        for test_file, ref_file in self.known_values:
#            result = codesort(test_file, True)
#            self.assertTrue(codesort.binary_file_compare(result, ref_file))
#            print(self.known_values)
            continue


class CodeBlockKnownValues(unittest.TestCase):
    """ Unit Testing for the CodeBlock class"""
    chunk_1 = """class HelloKitty(object):
    def __init__(self):
        pass

    def method(self):
        return 5"""

    chunk_2 = """def my_function(a, b):
        return a + b"""

    # (chunk, (name, type, num_lines))
    known_values = ((chunk_1, ("HelloKitty", 'class', 6)),
                    (chunk_2, ("my_function", 'function', 2)),
                    )

    def test_kvt_name(self):
        """ Known-value testing for the CodeBlock name attribute. """
        for chunk, expected in self.known_values:
            result_obj = codesort.CodeBlock(chunk)
            self.assertEqual(expected[0], result_obj.name)

    def test_kvt_type(self):
        """ Known-value testing for the CodeBlock type attribute. """
        for chunk, expected in self.known_values:
            result_obj = codesort.CodeBlock(chunk)
            self.assertEqual(expected[1], result_obj.code_type)

    def test_kvt_line_count(self):
        """ Known-value testing for the CodeBlock line_count attribute. """
        for chunk, expected in self.known_values:
            result_obj = codesort.CodeBlock(chunk)
            self.assertEqual(expected[2], result_obj.line_count)


class ClassifyBlock(unittest.TestCase):
    """ Unit testing for the classify_block function """
    ex1 = ('class Apple(object):', 'class')
    ex2 = ('   def my_func(a, b):', 'function')
    ex3 = ('y = 4x', 'instance_var')
    ex4 = ("""    @decorator1
    @decorator2
    def my_decorated_function(b, c):)""", 'function')
    ex5 = ('# a comment', 'comment')
    ex6 = ('from module import thing', 'import')
    ex7 = ('import other_module', 'import')
    ex8 = ('""" A docstring """', 'docstring')
    ex9 = ('CONSTANT = 57', 'constant')

    examples = [ex1,
                ex2,
                ex3,
                ex4,
                ex5,
                ex6,
                ex7,
                ex8,
                ex9,
                ]

    def test_known_values(self):
        """ Known-value testing for the classify_block function """
        for block, expected in self.examples:
            code_type = codesort.classify_block(block)
            self.assertEqual(code_type, expected)

    def test_constant_match(self):
        """ Makes sure that constants that match PEP8 standards will work """
        blocks = ['CONSTANT = "helloooo, nurse!"',
                  'SOME_CONSTANT = 57',
                  '99_PROBLEMS = "but a bug aint one!"',
                  'CONST_1 = 2',
                  ]

        for block in blocks:
            self.assertEqual(codesort.classify_block(block), 'constant')

    def test_constant_not_match(self):
        """ Test that some common invalid constants are not matched """
        blocks = ['Constant=[x for x in range(5)]',
                  'con=2',
                  'AAA=-11',
                  ]

        for block in blocks:
            self.assertNotEqual(codesort.classify_block(block), 'constant')

if __name__ == "__main__":
    unittest.main(exit=False, verbosity=2)


codesort/tests/test_ffp.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jul 02 17:06:51 2014

@name:          test_ffp.py
@vers:          0.1
@author:        dthor
@created:       Wed Jul 02 17:06:51 2014
@modified:      Wed Jul 02 17:06:51 2014
@descr:         Unit Testing for codesort.find_fold_points module
"""

from __future__ import print_function
import unittest
import os
import codesort.find_fold_points as ffp


class FindFoldPoints(unittest.TestCase):
    """ Test the find_fold_points function """
    root_dir = os.path.split(__file__)[0]
    test_data_path = r"test_data"
    file_1 = "sorted_1.py"
    file_2 = "2_multiline_defs.py"
    file_3 = "3_comments.py"
    file_1_path = os.path.join(root_dir, test_data_path, file_1)
    file_2_path = os.path.join(root_dir, test_data_path, file_2)
    file_3_path = os.path.join(root_dir, test_data_path, file_3)

    # Manually determined the (start, end, indent) values for the file
    file_1_result = {(10, 34, 1),
                     (12, 14, 2),
                     (16, 18, 2),
                     (20, 22, 2),
                     (24, 26, 2),
                     (28, 30, 2),
                     (32, 34, 2),
                     (37, 61, 1),
                     (39, 41, 2),
                     (43, 45, 2),
                     (47, 49, 2),
                     (51, 53, 2),
                     (55, 57, 2),
                     (59, 61, 2),
                     (64, 66, 1),
                     (69, 71, 1),
                     (74, 76, 1),
                     (79, 81, 1),
                     }

    file_2_result = {(10, 40, 1),
                     (14, 16, 2),
                     (22, 24, 2),
                     (30, 32, 2),
                     (38, 40, 2),
                     }

    file_3_result = {(10, 35, 1),
                     (14, 16, 2),
                     (22, 24, 2),
                     (26, 29, 2),
                     (31, 35, 2),
                     }

    test1 = ("""def myfunc(a, b):
    if a < 0:
        return b
    if a > 0:
        return a + b
    return 0

def myfunc2(a):
    return 'xsdfsd' + (a-2)*(a+3)""", {(1, 6, 1),
                                       (2, 3, 2),
                                       (4, 5, 2),
                                       (8, 9, 1),
                                       })

    test2 = ("""def func1(a):
    return a

def func2(b):
    return b

def func3(c):
    return c""", {(1, 2, 1),
                  (4, 5, 1),
                  (7, 8, 1),
                  })

    test3 = ("""import apples

def myfunc(a, b):
    if a < 0:
        return b
    if a > 0:
        return a + b
    return 0

def myfunc2(a):
    return 'xsdfsd' + (a-2)*(a+3)""", {(3, 8, 1),
                                       (4, 5, 2),
                                       (6, 7, 2),
                                       (10, 11, 1),
                                       })
    known_values = (test1,
                    test2,
                    test3,
                    )

    file_1_name = "sorted_1.py"
    file_1 = os.path.join(root_dir, test_data_path, file_1_name)
    file_1_result = {(10, 34, 1),
                     (12, 14, 2),
                     (16, 18, 2),
                     (20, 22, 2),
                     (24, 26, 2),
                     (28, 30, 2),
                     (32, 34, 2),
                     (37, 61, 1),
                     (39, 41, 2),
                     (43, 45, 2),
                     (47, 49, 2),
                     (51, 53, 2),
                     (55, 57, 2),
                     (59, 61, 2),
                     (64, 66, 1),
                     (69, 71, 1),
                     (74, 76, 1),
                     (79, 81, 1)}

    def test_known_values(self):
        """ docstring """
        for code_block, folds in self.known_values:
            result = ffp.find_fold_points(code_block)
            self.assertSetEqual(set(result), folds)

    def test_known_file(self):
        with open(self.file_1) as openfile:
            file_text = "".join(openfile.readlines())
        result = ffp.find_fold_points(file_text)
        self.assertSetEqual(set(result), self.file_1_result)

    def test_known_file2(self):
        """ Run find_fold_points a on known file. """
        with open(self.file_1_path) as openfile:
            file_text = "".join(openfile.readlines())
        result = ffp.find_fold_points(file_text)
        self.assertSetEqual(set(result), self.file_1_result)

    def test_multiline_defs(self):
        """ Make sure multiline definitions are folded after the : """
        with open(self.file_2_path) as openfile:
            file_text = "".join(openfile.readlines())
        result = ffp.find_fold_points(file_text)
        self.assertSetEqual(set(result), self.file_2_result)

    def test_comment_after_def(self):
        """ Check that comments after fold start are properly folded """
        with open(self.file_3_path) as openfile:
            file_text = "".join(openfile.readlines())
        result = ffp.find_fold_points(file_text)
        self.assertSetEqual(set(result), self.file_3_result)


def main():
    """ Main Code """
    unittest.main(exit=False, verbosity=1)


if __name__ == "__main__":
    main()


tests/test_data/sorted_1.py
# -*- coding: utf-8 -*-
"""
Docstring!
"""


from __future__ import print_function, division


def ClassA(object):
    """ ClassA, for sorting! """
    def __init__(self):
        """ Nothing needed here! """
        pass

    def _private_a(self):
        """ And things! """
        pass

    def _private_b(self):
        """ Stuff! """
        pass

    def public_a(self):
        """ I'm tired... """
        pass

    def public_b(self):
        """ That's all for now """
        pass

    def public_c(self):
        """ More stuff! """
        pass


class ClassB(object):
    """ ClassA, for sorting! """
    def __init__(self):
        """ Nothing needed here! """
        pass

    def _private_a(self):
        """ And things! """
        pass

    def _private_b(self):
        """ Stuff! """
        pass

    def public_a(self):
        """ I'm tired... """
        pass

    def public_b(self):
        """ That's all for now """
        pass

    def public_c(self):
        """ More stuff! """
        pass


def module_func_a(a):
    """ I hope this works """
    pass


def module_func_b(a, b):
    """ Please don't hate me """
    pass


def module_func_c(c):
    """ But in the original file, it's not! """
    pass


if __name__ == "__main__":
    """ This should be last """
    pass


# A comment that starts the orphaned module code
x = 5
y = 27
print(x+y)


codesort_proj/codesort/find_fold_points.py
# -*- coding: utf-8 -*-
"""
@name:          find_fold_points.py
@vers:          0.1
@author:        Douglas Thor
@created:       Sun Jun 29 17:03:12 2014
@modified:      Sun Jun 29 17:03:12 2014
@descr:         Returns the fold points - where code gets indented and
                dedented - of a .py file.
"""

from __future__ import print_function, division
import os.path
import tokenize
from StringIO import StringIO


def find_fold_points(block):
    """
    Returns a list of (start_row, end_row, indent) tuples that denote fold
    locations. Basically anywhere that there's an indent.
    """
    token_whitelist = (tokenize.NL,
                       tokenize.NEWLINE,
                       tokenize.INDENT,
                       tokenize.DEDENT,
                       tokenize.COMMENT,
                       )

    # temporary code that allows for running a block or a full file
    if os.path.isfile(block):
        with open(block) as open_file:
            token_block = tokenize.generate_tokens(open_file)
    else:
        token_block = tokenize.generate_tokens(StringIO(block).readline)

    indent_level = 0
    nl_counter = 0
    comment_counter = 0
    indents = []
    result = []
    for toknum, _, srowcol, _, _ in token_block:
        # Account for comments at the start of a block and newlines at the
        # end of a block.
        if toknum == tokenize.NL:
            nl_counter += 1
        if toknum == tokenize.COMMENT:
            comment_counter += 1
        if toknum == tokenize.INDENT:
            indent_level += 1
            indents.append(srowcol[0] - 1 - comment_counter)
        if toknum == tokenize.DEDENT:
            # the next DEDENT belongs to the most recent INDENT, so we pop off
            # the last indent from the stack
            indent_level -= 1
            matched_indent = indents.pop()
            result.append((matched_indent,
                           srowcol[0] - 1 - nl_counter,
                           indent_level + 1))
        if toknum not in token_whitelist:
            nl_counter = 0
            comment_counter = 0

    if len(indents) != 0:
        raise ValueError("Number of DEDENTs does not match number of INDENTs.")

    return result


if __name__ == "__main__":
    pass


codesort/codesort.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 18 16:50:37 2014

@name:          codesort.py
@vers:          0.1.0
@author:        dthor
@created:       Wed Jun 18 16:50:37 2014
@descr:         A python source code organization tool.

    Sorts the python file according to the sorting schema. Since class
    and function definitions do not need to be in execution order or
    definition-use order, we can organize the source code alphabetically.
    This tool will take a .py file and organize it. This helps the programmer
    maintain uniformity across his projects, making it easier to find
    function definitions.

    The Sorting Schema:
    1. Source file encoding ("# -- coding: utf-8 --")
    2. Module docstring / comments
    3. Import statements
        a. __future__ import statements
        b. standard library imports
        c. 3rd party imports
        d. local package imports
    4. Module constants
        a. "Magic" globals (__author__, __version__, etc.)
        b. Module globals
    5. Classes (organized alphabetically)
        a. Class docstring
        b. Overriding methods
            1. __init__
            2. __new__
            3. __del__
            4. __str__
            5. __repr__
            6. __cmp__
            7. __hash__
            8. __nonzero__
            9. __unicode__
            10. __getattr__
            11. __setattr__
            12. __delattr__
            13. __getattribute__
            14. __get__
            15. __set__
            16. __delete__
        c. Private methods (organized alphabetically)
        d. Public methods (organized alphabetically)
    6. Module functions (organized alphabetically)
    7. Main() function
    8. toplevel module code
    9. if __name__ == '__main__' block

Requires:
    docopt >= 0.6.1         Python command-line argument parser

Usage:
    codesort.py
    codesort.py FILE
    codesort.py FILE [-n]

Options:
    -n --new-file       # Create a new file rather than replacing the old one.
    -h --help           # Show this screen.
    --version           # Show version.
"""


from __future__ import print_function, division
from docopt import docopt
import os.path
import re
import tokenize
import StringIO
import find_fold_points as ffp


__author__ = "Douglas Thor"
__version__ = "CodeSort v0.1.0"
__license__ = "MIT"


BLOCK_TYPE = [(re.compile(r'[ ]*@.'), 'decorator'),
              (re.compile(r'[ ]*class .'), 'class'),
              (re.compile(r'[ ]*def .'), 'function'),
              (re.compile(r'[ ]*#.'), 'comment'),
              (re.compile(r'[ ]*from .'), 'import'),
              (re.compile(r'[ ]*import .'), 'import'),
              (re.compile(r'[ ]*""".'), 'docstring'),
              (re.compile(r'[ ]*[A-Z0-9_]+ = .*'), 'constant'),
              (re.compile(r'[ ]*[a-z0-9_]+ = .*'), 'instance_var'),
              (re.compile(r'[ ]*.'), 'other'),
              ]


class CodeSort(object):
    """
    Main CodeSort class

    Contains all the methods and attributes for the module.
    """
    def __init__(self, filepath):
        """ Init class attributes """
        self.filepath = filepath

    def __str__(self):
        """ String representation """
        return "CodeSort Class for {}".format(self.filepath)

    def sort(self):
        """
        Sorts the python file according to the sorting schema.
        """
#
#        The Algorithm:
#        First, we obviously need to open the file.
#        Then we collect information on the classes, functions, global
#        variables, etc.
#        Each code block should be put into some kind of container
#        This container should have an attribute for the code block name
#        so that I can use that for sorting.
#        After parsing all code blocks into the container, sort them and
#        re-write them into a new file. Always into a new file. Overwriting
#        the old file would be a 2nd step.
#        If the user wants a new file created, then we're done.
#        If the user wants the original file overwritten (default), then
#        we open original file, delete all contents, and then copy all
#        contents from the new file into the original. I intentionally do not
#        want to just delete the original and rename the new because that
#        could mess up some SVN to backup functions (because the file
#        creation date would change, or perhaps there's a file UID that would
#        change).
        pass


class CodeBlock(object):
    """
    Class that contains info of a given code block. A code block will be
    from the decorator (if any) to the end of the indent.


    Public Methods:BLOCK_TYPE
        I can't think of any right now...

    Private Methods:
        count the number of lines
        set the type
        set the name
        run the init
    """

    def __init__(self, code_text):
        """ Init class attributes """
        self.code_text = code_text
        self.code_type = ''
        self.line_count = 0
        self.decorator_count = 0
        self.name = ""
        self._init_attributes()

    def __str__(self):
        """ String representation """
        str_rep = "A {block_type} named {name} with {num} lines."
        return str_rep.format(block_type=self.code_type,
                              name=self.name,
                              num=self.line_count,
                              )

    def _set_line_count(self):
        """ Private method that actually sets the line_count attribute """
        self.line_count = len(self.code_text.splitlines())

    def _set_type(self):
        """ Private methon that actually sets the type attribute """
        self.code_type = classify_block(self.code_text)

    def _set_name(self):
        """ Private methon that actually sets the name of the code block """
        # Simple naming, where the line is a func or class
        name = self.code_text.lstrip().split(' ')[1].split('(')[0]
        self.name = name

    def _set_decorator_count(self):
        """ Counts the number of decorators """
        # TODO: add decorator counting algorithm.
        count = 0
        self.decorator_count = count

    def _init_attributes(self):
        """ Runs all the various parsers """
        self._set_line_count()
        self._set_type()
        self._set_name()
        self._set_decorator_count()


def file_prompt():
    """ Prompt the user for the file path """
    exit_set = {'exit', 'e', 'quit', 'q'}
    print("Please enter a file to convert or parse:")
    while True:
        try:
            filepath = raw_input("File: ")
            if filepath.lower() in exit_set:
                filepath = 'exit'
                break
            elif os.path.isfile(filepath):
                break
            else:
                raise ValueError
        except ValueError:
            print("Invalid file. Please enter the path again.")
    return filepath


def classify_block(code_block):
    """ Classifies a code block to one BLOCK_TYPE category """
    match = False

    for regex_key, code_type in BLOCK_TYPE:
        if regex_key.match(code_block):
            # if it's a decorator, we need to go deeper
            if code_type == 'decorator':
                new_block = '\n'.join(code_block.split('\n')[1:])
                code_type = classify_block(new_block)
            return code_type

    if not match:
        return 'unknown'


def binary_file_compare(file1, file2):
    """
    Compares two files byte-by-byte. Usefull if the md5sum is different
    for some reason.
    """
    match = False
    with open(file1, 'rb') as ref:
        with open(file2, 'rb') as tmp:
            # File Size Check
            ref.seek(-1, 2)
            tmp.seek(-1, 2)
            ref_size = ref.tell()
            if not ref_size == tmp.tell():
                return match
            # Last byte check
            if not ref.read() == tmp.read():
                return match
            # move back to the begining and iterate through the file
            ref.seek(0, 0)
            tmp.seek(0, 0)
            for ref_byte, tmp_byte in zip(ref, tmp):
                match = ref_byte == tmp_byte
                if not match:
                    break
    return match


def print_tokens(code):
    """ Prints out the tokenized form of code in an easy-to-ready format. """
    import os.path

    if os.path.isfile(code):
        with open(code) as open_file:
            text = ''.join(open_file.readlines())
    else:
        text = code

    token_text = tokenize.generate_tokens(StringIO.StringIO(text).readline)

    indent_pos = 0
    format_str = "{num:>3}\t{name:<10}\t{indent}\t{st}\t{end}\t|{line:<30}"
    for toknum, tokval, srowcol, erowcol, logical_l in token_text:
        print(format_str.format(num=toknum,
                                name=tokval.strip()[:7].replace("\n", "/n"),
                                st=srowcol,
                                end=erowcol,
                                line=logical_l.rstrip()[:30].replace("\n",
                                                                     "/n"),
                                indent=indent_pos,
                                ))

        if toknum == tokenize.INDENT:
            indent_pos += 1
        if toknum == tokenize.DEDENT:
            indent_pos -= 1


def find_fold_points(block):
    """
    Returns a list of (start_row, end_row, indent) tuples that denote fold
    locations. Basically anywhere that there's an indent.
    """
    return ffp.find_fold_points(block)


def split_blocks(code, fold_points):
    """ splits code into separate blocks """
    # if a fold point starts with class or def, then it's the start of a block
    code_lines = code.splitlines()
    for start, end, indent in fold_points:
        if code_lines[start - 1].lstrip().startswith('def'):
            block = '\n'.join(code_lines[start - 1:end])
            print("----------")
            print(block)
            print("----------")
    return


def main():
    """ Main Code """
    args = docopt(__doc__, version=__version__)

#    if args['FILE'] is None:
#        args['FILE'] = file_prompt()
#        if args['FILE'] == 'exit':
#            print("Exiting Program")
#            return

    if args['--new-file']:
        print("A new file will be made.")

# ---------------------------------------------------------
# Quick Testing
# ---------------------------------------------------------

    cs = CodeSort(args['FILE'])
    print(cs)

    block_1 = """class HelloKitty(object):

    def __init__(self):
        if bagle:
            eat
        else:
            sleep

    def method(self):
        return 5"""

    block_2 = """def my_function(a, b):
    'string'
    return a + b

def my_func_b(x):
    return x

print(5)"""

    # (block, (name, type, num_lines))
    known_values = ((block_1, ("HelloKitty", 'class', 6)),
                    (block_2, ("my_function", 'function', 2)),
                    )

    # KVT for the CodeBlock class.
    for block, expected in known_values:
        result_obj = CodeBlock(block)
        print(result_obj)

    print()
    print()
# ---------------------------------------------------------
# End Quick Testing
# ---------------------------------------------------------
#    print(find_fold_points(block_1))

    root_dir = os.getcwd()
    test_data_path = r"tests\\test_data"
    file_1_name = "sorted_1.py"
    file_1 = os.path.join(root_dir, test_data_path, file_1_name)
#    print_tokens(file_1)
    
#    folds = find_fold_points(block_2)
#    print(folds)
#    split_blocks(block_2, folds)
    


if __name__ == "__main__":
    main()
